Add simple time slider to map view

DavidMStraub opened the pull request:
Hi @1ec5, I added a simple time range slider to the map - it doesn't yet do anything, except setting properties on the map view. I can add filtering on the map pins later. How would we now filter the OHM layer?

DavidMStraub committed:
Add simple time slider to map view

1ec5 commented on src/views/GrampsjsViewMap.js:
```diff
@@ -154,6 +163,11 @@ export class GrampsjsViewMap extends GrampsjsView {
   }
 
+  _handleTimeSliderChange(event) {
+    this._dateRangeStart = event.detail.start
+    this._dateRangeEnd = event.detail.end
```
It looks like this code would have access to the Leaflet map via `map._map`. From there, it’s another hop to get to the MapLibre `Map` object, which can set filters and such.

Here’s some public domain code for filtering an OHM style by date tags:

https://github.com/OpenHistoricalMap/ohm-embed-1yr/blob/d8d7f2362bfd87521c4559cf59885a1282e52621/embed-map-single-year.js#L77-L170

At some point, we should package it up as a proper library, independent of any time slider UI, but for now it’s stable and straightforward enough to copy-paste around.

1ec5 replied:
OpenHistoricalMap/issues#840

1ec5 reviewed (commented):
Thanks so much for putting this together – range sliders are very cool and we need to use them more in OHM too. I do wonder if a single-knob slider would be more logical in some of the map views, where we don’t overlay a bunch of pins that you might want to filter. But we can figure that out separately.

DavidMStraub commented:
>  I do wonder if a single-knob slider would be more logical in some of the map views, where we don’t overlay a bunch of pins that you might want to filter. 

Yes indeed. Perhaps a single-knob slider and a button/dropdown to select a range of &pm;5, &pm;10, &pm;50, &pm;100 years around the selected date to filter the pins?

1ec5 commented on src/components/GrampsjsMapTimeSlider.js:
```diff
@@ -0,0 +1,58 @@
+    return html`
+      <div id="container">
+        <md-slider
+          @change="${this._handleInput}"
```
Would it be possible to fire this event continuously as the user scrubs the slider? The user might find it confusing at first if the map doesn’t seem to update until they let go of the knob.

DavidMStraub replied:
True. We can just change it to `@input`.

DavidMStraub committed:
Use single-knob slider, add map filter

DavidMStraub commented:
I changed it to a normal slider now and added the filter methods; however it doesn't seem to work yet; or it works only once, to be precise. Not sure what I'm doing wrong.

1ec5 commented on src/components/GrampsjsMapTimeSlider.js:
```diff
@@ -0,0 +1,88 @@
+
+  constructor() {
+    super()
+    this.value = 2024
```
This would calculate the initial date dynamically, a century before present so the OSM and OHM layers look visually distinct:

```suggestion
    this.value = new Date().getFullYear() - 100
```

Though maybe defaulting to the present would keep things simpler.

1ec5 commented on src/components/GrampsjsMapTimeSlider.js:
```diff
@@ -0,0 +1,88 @@
+          @change="${this._handleInput}"
+          labeled
+          min="1500"
+          max="2024"
```
Calculate the maximum date dynamically:

```suggestion
          max="${new Date().getFullYear()}"
```

1ec5 commented on src/util.js:
```diff
@@ -588,3 +588,120 @@ export function dateIsEmpty(date) {
+  map.getStyle().layers.map(layer => {
+    if (!('source-layer' in layer)) return
+
+    const filter = constrainFilterByDate(layer.filter, decimalYear)
```
> I changed it to a normal slider now and added the filter methods; however it doesn't seem to work yet; or it works only once, to be precise. Not sure what I'm doing wrong.

Ah, the original embed code was only designed to filter the map initially without any updates. OpenHistoricalMap/ohm-embed-1yr#5 will fix this issue in order to implement animation in that embed. The relevant part of that PR is changing this line to get the most up-to-date filter, rather than the one that happens to have been in the layer the last time we got the style:

```suggestion
    const filter = constrainFilterByDate(map.getFilter(layer.id), decimalYear)
```

1ec5 commented on src/util.js:
```diff
@@ -588,3 +588,120 @@ export function dateIsEmpty(date) {
+  if (filter && filter[0] === 'all' && filter[1] && filter[1][0] === 'any') {
+    if (
+      filter[1][2] &&
+      filter[1][2][0] === '<=' &&
+      filter[1][2][1] === 'start_decdate'
+    ) {
+      newFilter[1][2][2] = decimalYear
+    }
+    if (
+      newFilter[2][2] &&
+      newFilter[2][2][0] === '>=' &&
+      newFilter[2][2][1] === 'end_decdate'
+    ) {
+      newFilter[2][2][2] = decimalYear
+    }
+    return newFilter
+  }
+
+  const dateFilter = [
+    'all',
+    ['any', ['!has', 'start_decdate'], ['<=', 'start_decdate', decimalYear]],
```
This filter is currently designed to show only the features that existed as of the very moment that is passed in. Since we’re passing in a whole number, it’s only going to show what existed as of January 1 of the year. Would that be alright, or should we loosen the filter to show anything that existed at any time within the given year? The downside is that sometimes you’ll get overlapping features, like two different shops in the same place (because one replaced the other partway through the year) or a construction area on top of the finished construction (because construction finished partway through the year).

```suggestion
      filter[1][2][0] === '<' &&
      filter[1][2][1] === 'start_decdate'
    ) {
      newFilter[1][2][2] = decimalYear + 1
    }
    if (
      newFilter[2][2] &&
      newFilter[2][2][0] === '>=' &&
      newFilter[2][2][1] === 'end_decdate'
    ) {
      newFilter[2][2][2] = decimalYear
    }
    return newFilter
  }

  const dateFilter = [
    'all',
    ['any', ['!has', 'start_decdate'], ['<', 'start_decdate', decimalYear + 1]],
```

DavidMStraub replied:
I think showing anything that existed during the year makes indeed more sense in our case.

The problem with overlapping construction site & building would happen regardless, just shifted by one year, isn't it?

1ec5 replied:
The filter is currently written as a precise moment in time. We would show two overlapping features if they’re only tagged with year precision, or if they’re tagged with day precision but just happen to change over on New Year’s Day. It’s probably not a big deal, come to think of it. Someone who cares about this imprecision would want the ability to set a specific date anyways.

DavidMStraub committed:
Update src/util.js

Co-authored-by: Minh Nguyễn <mxn@1ec5.org>

DavidMStraub committed:
Update src/components/GrampsjsMapTimeSlider.js

Co-authored-by: Minh Nguyễn <mxn@1ec5.org>

DavidMStraub committed:
Update src/components/GrampsjsMapTimeSlider.js

Co-authored-by: Minh Nguyễn <mxn@1ec5.org>

DavidMStraub committed:
Use input event instead of change

1ec5 commented:
Thanks so much for figuring out a nice-looking UI for this. From what I can tell, the main things to sort out are:

* Show and hide the time slider based on the active layer
* How should the time slider filter pins? Would we show all the events up to the selected date? Or should time-based pin filtering be a separate control in the dialog box accessible from the funnel button in the search bar?

Do you want to keep working on this branch or merge into gramps-project/gramps-web#454 and work on it there? (I enabled edits by maintainers so you can push commits directly to that branch.)

DavidMStraub committed:
Update src/util.js

Co-authored-by: Minh Nguyễn <mxn@1ec5.org>

DavidMStraub commented:
Agreed!

> Show and hide the time slider based on the active layer

I was thinking we could leave it present all the time if it's also used to filter pins. In the OSM case it would affect only pins, in the OHM case map & pins.

> How should the time slider filter pins? Would we show all the events up to the selected date? Or should time-based pin filtering be a separate control in the dialog box accessible from the funnel button in the search bar?

I added the 1924 &pm; 10 without function but I was thinking the &pm; x could be clickable to select a wider range. We could then show all the pins in that time period (e.g. in which places did my ancestors have events at in the 18th century).

> Do you want to keep working on this branch or merge into https://github.com/gramps-project/gramps-web/pull/454 and work on it there? (I enabled edits by maintainers so you can push commits directly to that branch.)

Yes, let's do it like that, as soon as the pin filtering is implemented we can then merge the main PR.

Edit: I meant yes, let's merge this into your `openhistoricalmap-437` feature branch.

1ec5 commented:
Sounds good to me!
